test_log/src/main/java/org/corfudb/test_log/TestLog.java

```diff
@@ -0,0 +1,289 @@
+                if (use_output_file) {
+                    FileOutputStream f_strm_o = new FileOutputStream(f_out);
+                    FileChannel f_chan_o = f_strm_o.getChannel();
+                    processLogFile(f_strm_i.getChannel(), f_chan_o, op);
+                    f_strm_o.close();
+                } else {
+                    processLogFile(f_strm_i.getChannel(), null, op);
+                }
+                f_strm_i.close();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+    private static void processLogFile(FileChannel fc_in,
+                                       FileChannel fc_out,
+                                       Operation op) throws IOException {
+        boolean display = op.getOpType() == Operation.OperationType.DISPLAY ||
+                    op.getOpType() == Operation.OperationType.DISPLAY_ALL;
+        boolean display_data = op.getOpType() == Operation.OperationType.DISPLAY_ALL;
+
+        System.out.format("processLogFile display %d output %d\n", display ? 1 : 0,
+                fc_out != null ? 1 : 0);
+
+        // Metadata
+        //   LogHeader
+        // Metadata
+        //   LogEntry
+        // ...
+        fc_in.position(0);
+        long file_size = fc_in.size();
+        ByteBuffer md_buffer = ByteBuffer.allocate(MetadataSize);
+        int r = fc_in.read(md_buffer);
+        md_buffer.flip();
+        if (fc_out != null) {
+            fc_out.write(md_buffer);
+        }
+        if (r > 0) {
+            Metadata md = Metadata.parseFrom(md_buffer.array());
+            if (display) {
+                System.out.println("checksum " + String.format("%08x", md.getChecksum()));
+                System.out.println("length " + Integer.toString(md.getLength()));
+            }
+            ByteBuffer lh_buffer = ByteBuffer.allocate(LogHeaderSize);
+            r = fc_in.read(lh_buffer);
+            lh_buffer.flip();
+            if (fc_out != null) {
+                fc_out.write(lh_buffer);
+            }
+            if (r > 0) {
+                LogHeader lh = LogHeader.parseFrom(lh_buffer.array());
+                if (display) {
+                    System.out.println("version " + Integer.toString(lh.getVersion()));
+                    System.out.println("verify " + Boolean.toString(lh.getVerifyChecksum()));
+                }
+            }
+        }
+        HashFunction hf = Hashing.crc32c();
+        long rem_size = fc_in.size() - fc_in.position();  // if size == position then file pointer is off the end
+        int record_cnt = 0;
+        boolean replace_entry = false;
+        boolean enable_write = false;
+        switch (op.getOpType()) {
+            case ERASE_ONE:
+            case ERASE_REMAINING:
+                enable_write = true;
+                break;
+            default:
+                enable_write = false;
+        }
+        while (rem_size > 0) {
+            boolean trigger = false;
```

Comment:
1. perhaps we could modularize here a bit, e.g., a method for reading and parsing individual entries

2. some comments would be nice :) ..